Timestamp in recipient

fboucquez opened the pull request:
Fixes https://github.com/symbol/catapult-rest/issues/636

It adds timestamp next to the id in transaction statements.

![image](https://user-images.githubusercontent.com/5390558/129585014-2dcd3515-b4b1-4e4e-8b94-636bd95f0851.png)

This implementation runs a second query looking for blocks for the given block heights. It does not use aggregate queries. I tried but failed to create an aggregate query between the `transactionStatements` and `blocks` collections.

If a mongo aggregate query is desired, I can try again to make it work. Note that the aggregate is between nested properties, not top-level properties.

fboucquez commented:
Once this is approved, the solution can be replicated for https://github.com/symbol/catapult-rest/issues/637

Jaguar0625 commented on catapult-sdk/test/plugins/receipts_spec.js:
```diff
@@ -59,7 +59,7 @@ describe('receipts plugin', () => {
 			['addressResolution', 'mosaicResolution', 'transaction'].forEach(statementType => {
-				expect(Object.keys(modelSchema[`${statementType}Statement`]).length).to.equal(2);
+				expect(Object.keys(modelSchema[`${statementType}Statement`]).length).to.equal(3);
 				expect(modelSchema[`${statementType}Statement`]).to.contain.all.keys(['statement']);
```
you might want to update this check

fboucquez replied:
Ok, I see, thanks. Would we want to add timestamp to `addressResolutionStaments` and `mosaicResolutionStatements`? 

To these endpoints?

http://ngl-dual-101.testnet.symboldev.network:3000/statements/resolutions/address
http://ngl-dual-101.testnet.symboldev.network:3000/statements/resolutions/mosaic

Or separate transactionStements from the other 2 is fine?

Jaguar0625 replied:
either is fine

fboucquez replied:
https://github.com/symbol/symbol-openapi/issues/279 for the open api change

Jaguar0625 commented on rest/test/plugins/receipts/ReceiptsDb_spec.js:
```diff
@@ -181,12 +185,12 @@ describe('receipts db', () => {
-					createTransactionStatement(20, 110, [
+					createTransactionStatement(20, 100, [
 						createReceipt({ recipientAddress: testAddress1 }),
 						createReceipt({ recipientAddress: testAddress2 })
 					]),
-					createTransactionStatement(30, 120, [createReceipt({ recipientAddress: testAddress2 })]),
-					createTransactionStatement(40, 130, [])
+					createTransactionStatement(30, 100, [createReceipt({ recipientAddress: testAddress2 })]),
+					createTransactionStatement(40, 100, [])
```
why did you need to change these?

fboucquez replied:
I was testing the unique ids, it's tested elsewhere so rolling this back

Jaguar0625 commented on rest/test/db/utils/dbTestUtils.js:
```diff
@@ -269,6 +269,17 @@ const runDbTest = (dbEntities, collectionName, createDbFacade, issueDbCommand, a
 		.then(() => db.close());
 };
 
+const insertItems = (dbEntities, collectionName) => {
```
i think insertEntities would be more consistent with other naming in this file

Jaguar0625 commented on rest/src/db/CatapultDb.js:
```diff
@@ -225,6 +225,17 @@ class CatapultDb {
+		return this.queryDocumentsAndCopyIds(
+			'blocks',
+			{ 'block.height': { $in: heights } },
+			{ projection: { 'meta.transactionMerkleTree': 0, 'meta.statementMerkleTree': 0 } }
```
why this projection?

Jaguar0625 replied:
don't you only need timestamp?

fboucquez replied:
> why this projection?

To keep it consistent with the singular `blockAtHeight` 
https://github.com/symbol/catapult-rest/blob/e730a5ea4d8c25fc4aedf534962de4040eef3c83/rest/src/db/CatapultDb.js#L220. 

> don't you only need timestamp?

In this case yes, for #637 fix I would need the feeMultipler. 

Having said that, I'll add the custom projection.

fboucquez replied:
Added custom projection + unit tests

Jaguar0625 commented on rest/test/db/CatapultDb_spec.js:
```diff
@@ -443,6 +443,42 @@ describe('catapult db', () => {
 			);
 		});
+
+		it('can retrieve blocks with transactions at heights', () => {
```
consider grouping these in own describe

Jaguar0625 replied:
consider swapping order of tests so they go from simpler to more complex

fboucquez replied:
Added describe, a new projection unit test and resorted tests.

fboucquez commented:
@Jaguar0625 , do you like the location of the timestamp next to id? I'm not sure where is the best place, top-level timestamp, inside statement, in a meta object/field, or in a (partial) block object/field.

Jaguar0625 commented:
> @Jaguar0625 , do you like the location of the timestamp next to id? I'm not sure where is the best place, top-level timestamp, inside statement, in a meta object/field, or in a (partial) block object/field.

i think in meta would probably be preferred. @gimre-xymcity do you have any opinion?

fboucquez commented:
Remaining questions:
1 - Where to put the timestamp in the response?
2 - Should we add the timestamp to mosaic and address statements for consistency?

gimre-xymcity commented:
> > @Jaguar0625 , do you like the location of the timestamp next to id? I'm not sure where is the best place, top-level timestamp, inside statement, in a meta object/field, or in a (partial) block object/field.
> 
> i think in meta would probably be preferred. @gimre-xymcity do you have any opinion?

meta is probably most appropriate

fboucquez commented:
@Jaguar0625 , I've moved the timestamp to the meta field and I have added the meta/timestamp to address and mosaic statements. 

![image](https://user-images.githubusercontent.com/5390558/129890271-4e8e5094-6c51-4b2e-87e5-14c33fc87cc9.png)

fboucquez commented on rest/src/plugins/receipts/ReceiptsDb.js:
```diff
@@ -109,7 +111,41 @@ class ReceiptsDb {
+		);
+		const blocks = await this.catapultDb.blocksAtHeights(blockHeights,
+			{ 'block.timestamp': 1, 'block.height': 1 });
+		const data = page.data.map(pageItem => {
+			const statementBlock = blocks.find(
+				blockInfo => blockInfo.block.height.equals(
+					pageItem.statement.height
+				)
+			);
+			if (!statementBlock) {
```
Thanks for the explanation @Jaguar0625 . 

I considered the index map but my thinking was, the max page size is 100 (default is 20), in general, the worst case would be 100 individual blocks. The height field is a Long object, I cannot index by height as-is. I would need to create a Map<string, Block> where the index/key is a string representation of the Long (probably decimal of hex representation). The cost of converting the heights to string (to create the map and when looking by height index) could be worse than just the O(nlog(n)) of 100 elements, plus the memory cost of the temp map.

Do you have a better way of doing the index map? I can change it if you like.

Jaguar0625 replied:
up to @gimre-xymcity

fboucquez replied:
@gimre-xymcity , are happy about the answer? Do you want the list to be indexed?

gimre-xymcity replied:
don't mind

fboucquez commented on rest/test/db/CatapultDb_spec.js:
```diff
@@ -445,6 +443,72 @@ describe('catapult db', () => {
 		});
 	});
 
+	describe('', () => {
```
Uff, missed that. I'll add

fboucquez commented on rest/src/plugins/receipts/ReceiptsDb.js:
```diff
@@ -109,7 +111,41 @@ class ReceiptsDb {
+					`Cannot find timestamp in block with height ${pageItem.statement.height.toString()}`
+				);
+			}
+			return { meta: { timestamp: statementBlock.block.timestamp }, ...pageItem };
```
Although it's not expected the statements to have meta right now, I've added the original meta + a comment clarifying this.
